### PR TITLE
Add CodeQL Analysis workflow

### PR DESCRIPTION
**Pull request by jhutchings1**
Adds a GitHub Actions workflow that runs CodeQL on every push, and on a daily schedule.

Code scanning looks for vulnerabilities, such as XSS, SQL injection, etc., in your code. If it finds any new vulnerabilities it surfaces them in the PR as check annotations, and blocks the build until they’re fixed or marked as false positives. If it finds any on the repo’s default branch it displays them in the [security tab](../security/code-scanning).

For now you also need to be feature flagged individually to see results in the security tab (as well as having write permission on this repo) - if you drop an email to jhutchings1@github.com I can get anyone you need added. 

Finally, this is an early access program that has not been released yet, so please don't share before May 6th when we’re unveiling it at GitHub Satellite.

Cc: @greysteil

### CODE REVIEW

**Comment by nicksnyder** [comment]
> blocks the build until they’re fixed or marked as false positives

Can you clarify the behavior here? I am interested to know if this is treated at status check that is either red/green. If so, do we have control over whether it blocks merging PRs or not (like other status checks)?

How does a developer mark something as a false positive? Are there docs somewhere I can read?

> For now you also need to be feature flagged individually to see results in the security tab (as well as having write permission on this repo) - if you drop an email to jhutchings1@github.com I can get anyone you need added.

Is it possible to seed the whitelist with everyone in our org? https://github.com/orgs/sourcegraph/people

**Comment by nicksnyder** [comment]
@jhutchings1 Also, could you please sign our CLA, thanks!

**Comment by greysteil** [comment]
> I am interested to know if this is treated at status check that is either red/green. If so, do we have control over whether it blocks merging PRs or not (like other status checks)?

It's just a status check - nothing special about it :-)

> How does a developer mark something as a false positive? 

The alerts show up as annotations on the diff,  and there's a "close" button on them, from where anyone with write access can mark them as false positive or acceptable risk. Once they do it updates the status check for that run, too (so it will move from red to green).

> Is it possible to seed the whitelist with everyone in our org?

Yep! I'll get that done now.

> @jhutchings1 Also, could you please sign our CLA, thanks!

This one needs you, @jhutchings1 !

File: .github/workflows/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v1
+      # Override language selection by uncommenting this and choosing your languages
+      # with:
+      #   languages: go, javascript, csharp, python, cpp, java
```
Comment (unknwon) [inline .github/workflows/workflows/codeql.yml:27]:
```suggestion
      with:
        languages: go, javascript
```

Is `typescript` a thing here?

File: .github/workflows/workflows/codeql.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+name: "Code Scanning - Action"
+
+on:
+  push:
+  schedule:
+    - cron: '0 0 * * 0'
```
Comment (unknwon) [inline .github/workflows/workflows/codeql.yml:6]: qq: this action will run on every single push, so what extra benefit one would get by running it daily?

File: .github/workflows/workflows/codeql.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v1
+      # Override language selection by uncommenting this and choosing your languages
+      # with:
+      #   languages: go, javascript, csharp, python, cpp, java
+
+    # Autobuild attempts to build any compiled languages  (C/C++, C#, or Java).
+    # If this step fails, then you should remove it and run the build manually (see below).
+    - name: Autobuild
+      uses: github/codeql-action/autobuild@v1
```
Comment (unknwon) [inline .github/workflows/workflows/codeql.yml:32]: Just to confirm, "Autobuild" works for Go?

File: .github/workflows/workflows/codeql.yml (hunk 1 above)
Comment (greysteil) [reply .github/workflows/workflows/codeql.yml:27]: The Javascript analysis works with typescript :-)

File: .github/workflows/workflows/codeql.yml (hunk 2 above)
Comment (greysteil) [reply .github/workflows/workflows/codeql.yml:6]: The set of security queries we use gets better all the time. If you don't push very often (not the case on sourcegraph/sourcegraph!) then running once a week will sometimes surface new alerts due to query improvements

File: .github/workflows/workflows/codeql.yml (hunk 3 above)
Comment (greysteil) [reply .github/workflows/workflows/codeql.yml:32]: Sure does!

File: .github/workflows/workflows/codeql.yml (hunk 2 above)
Comment (unknwon) [reply .github/workflows/workflows/codeql.yml:6]: Gotcha, make sense! Weekly makes the most sense for sourcegraph/sourcegraph I think (ideally, Monday 00:00 but cc @nicksnyder here).

So if a new alert is surfaced, in which form that we will be notified (issue, PR, or)?

Comment (greysteil) [reply .github/workflows/workflows/codeql.yml:6]: It's be in the security tab on the repo. We don't currently send notifications for them but will add that in future.

The below will do weekly on Monday at 00:00 UTC (and the previous setup was for weekly - Justin misspoke in the PR text).

```suggestion
    - cron: '0 0 * * 1'
```

**Comment by jhutchings1** [comment]
I confirmed that the CodeQL analysis run is building properly now, so you're good to merge any time from my POV. https://github.com/jhutchings1/sourcegraph/actions/runs/94128082

**Comment by unknwon** [comment]
Thanks @jhutchings1!

I took a look at the https://github.com/jhutchings1/sourcegraph/actions/runs/94128082 and have few more questions:
1. It took 13m 32s to finish, is there a way to we can reduce the time, for example, by caching Go and JS's dependencies (or it does already)?
2. In **Extracting go** step, logs are prefixed with `[build-err]` (e.g. https://github.com/jhutchings1/sourcegraph/runs/639582274?check_suite_focus=true#step:5:20) which looks like should only occur when there are errors? However, in **Extracting javascript**, logs have `[build]`. Why are they being inconsistent here?

**Comment by jhutchings1** [comment]
@unknwon I'll try an experiment with dependency caching, and follow-up with the team on the build-err you're seeing.

**Comment by nicksnyder** [comment]
I am merging this to try it out. We can revert if there are any issues.

**Comment by nicksnyder** [comment]
Just FYI, this was too slow for us to run on every push, so we are going to update it to run hourly.
https://github.com/sourcegraph/sourcegraph/pull/10875

**Comment by marcogario** [comment]
:wave: I am part of the Github team working on this feature. Thanks for the feedback! 

I was wondering if you would be up to try to split the analysis job by language. This would basically reduce the analysis time by 50%. The easiest way to do this is to use a matrix build. The following workflow file achieves that:
```yml
name: "Code Scanning - Action"

on:
  pull_request:
    branches: [master]
  push:
    branches: [master]

jobs:
  CodeQL-Build:

    strategy:
      fail-fast: false
      matrix:
        languages: [ 'go', 'javascript']

    # CodeQL runs on ubuntu-latest, windows-latest, and macos-latest
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    # Initializes the CodeQL tools for scanning.
    - name: Initialize CodeQL
      uses: github/codeql-action/init@v1
      with:
        languages: ${{ matrix.languages }}

    # Autobuild attempts to build any compiled languages  (C/C++, C#, or Java).
    # If this step fails, then you should remove it and run the build manually (see below).
    - name: Autobuild
      uses: github/codeql-action/autobuild@v1

    - name: Perform CodeQL Analysis
      uses: github/codeql-action/analyze@v1
```

The key point is that we define the two languages that we want to analyze (go and javascript) in the `strategy` section, and then run the same job for each of the languages, but tell the codeql-action/init step to only analyze one language at the time. This yields two parallel jobs, each taking approximately 8-9 minutes to run.

image

Please note that changing the jobs in which the analysis is performed might close and reopen existing alerts, as   we have no way of knowing that these two jobs are the same as the previous unified one.

**Comment by nicksnyder** [comment]
Splitting the analysis by language seems like a smart thing to do regardless, but 8 minutes would still slow our CI on PRs (our current PR builds take 5 minutes) so I would be hesitant to enable it on PRs.

Our current master builds (which do things like build and push docker images) take 11 minutes, so we could consider doing these checks per commit there.

**Comment by nicksnyder** [comment]
@marcogario I made this change in https://github.com/sourcegraph/sourcegraph/pull/12441 but now it looks like we are seeing an orphaned action on every build. How can we get rid of this?

Screen Shot 2020-07-27 at 9 35 33 PM
https://github.com/sourcegraph/sourcegraph/runs/917230867

**Pull request merged**